_posts/2016-05-22-th1eadin0g1.markdown

```diff
@@ -0,0 +1,202 @@
+
+These provide the lowest level API to threads, everything is built on top of this.
+
+### "Same Time"
+
+Saying that something is happening at the same time, is potentially a lie, or more accurately a simplification. I'm currently writing this on a MacBook Pro, with these specs: 
+
+```
+Hardware Overview:
+
+  Model Name:	MacBook Pro
+  Model Identifier:	MacBookPro12,1
+  Processor Name:	Intel Core i7
+  Processor Speed:	3.1 GHz
+  Number of Processors:	1
+  Total Number of Cores:	2
+  L2 Cache (per Core):	256 KB
+  L3 Cache:	4 MB
+  Memory:	16 GB
+```
+
+Which means there's one processor with two cores. As a rule of thumb, a processor can only process one instruction at a time. However, a processor can fake being more than one processor at once. Which is essentially what cores mean. 
+
+Hyperthreading is a tool used to help improve this faking of multiple processors. There's no way am I going to describe that though, check out the crazy illustration on [it's wikipedia page](https://en.wikipedia.org/wiki/Hyper-threading).
+
+So you can have a bunch of `pthread` functions, and they can "kinda" run code concurrently.
+
+### Why Bother?
+
+Well, it turns out people want to do more than one thing at once. 
+
+It's pretty reasonable to expect to be able to scroll a user interface without waiting for an image to download from the server or file system before you see the next step in the animation. My first blog post on this blog was actually [about this back in 2012](https://artsy.github.io/blog/2012/09/13/on-grid-thumbnails/).
+
+> In our naiveté, Folio was originally using UIImage's initWithContentsOfFile: to load (without caching) a jpg from the file system. Once the file was loaded into memory, we displayed it onscreen in an UIImageView. This was fast enough to deal with our small thumbnails of 240x240 but the moment that you start asking it to pull 3 or 4 480x480 jpg files off the filesystem, decompress them and then put them on the screen, you're not going to have a smooth scroll.
+
+Loading a JPG from a file to the screen is _slow_, so we were blocking the main thread and then the scrolling animation could not run until all of the loading had finished. This is the perfect example of something that could be fixed with background threading.  So, we're going to go through different ways in which we could attack the problem.
+
+### In the beginning, there was `NSThread`
+
+As a battle-scarred Mac Objective-C developer, I started writing code before [Grand Central Dispatch](https://en.wikipedia.org/wiki/Grand_Central_Dispatch) so we're going to go over the axiomic `NSThread` class before we jump to the higher abstractions that you would use in modern code, though there are uses for `NSThread` still - we use it in [the CocoaPods App](https://github.com/CocoaPods/CocoaPods-app/blob/d2d62327906213b31bb8e07f9b45a7c11b564477/app/CPReflectionService/RBObject%2BCocoaPods.m#L156).
```

Comment:
We use `NSThread` in the CocoaPods app, because we need a _single specific_ thread. A GCD queue may perform tasks on various threads and is thus not suited for the job.

(The reason we need a single thread is because we run the Ruby runtime on that thread, it would go FUBAR if it were to be ran from multiple threads.)